### PR TITLE
Note generator app

### PR DESCRIPTION
**Pull request by Haryohmi**
## Description

I developed a Notes Generator Tool to generate concise, structured notes from various types of input in other to provide students with clear, summarized learning materials. 

## Related Issue

[https://github.com/orgs/marvelai-org/projects/8/views/1?pane=issue&itemId=89621248&issue=marvelai-org%7Cmarvel-ai-backend%7C129](url)

## Type of Change

- [ ] **New feature**: A non-breaking change that adds "Note generator" feature to the app.
- [ ] **Documentation update**: New feature code is well documented
- [ ] **Test enhancement**: Adding or updating tests; no production code change.

## Proposed Solution

1. Input Handling
    * Implementation:
        * Built logic to accept three types of input: plain text, file uploads (PLAIN TEXT, CSV, PDF, DOCX), and URLs (websites, YouTube.
       * Used helper functions (extract_text_from_file, extract_text_from_url) for modularity, making the code extensible for future input types.
    * Design Decisions:
        * Modularized input parsing to support easy integration of new file formats.
        * Used mock-based testing to isolate file and URL handlers.
2. Notes Generation Logic
    * Implementation:
        * Developed an executor function that coordinates input parsing, focus-based filtering, and AI-driven note summarization.
        * Used flexible output structures (bullet points, paragraphs, tables) to match user preferences.
    * Design Decisions:
        * Separated content preprocessing and notes generation for clarity and maintainability.
        * Added an optional "focus" parameter to tailor notes to specific topics, improving usability.
3. Output Formats
    * Implementation:
        * Created functions to export notes in plain text, DOCX, and PDF formats using libraries like docx and FPDF.
   * Design Decisions:
        * Chose widely used formats to ensure compatibility with users' needs.
        * Structured output generation modularly to enable quick addition of new export formats.
4. Core Architecture
    * Implementation:
        * Organized code into a structured folder hierarchy (app/features/notes_generator/) for maintainability.
       * Used core.py for primary logic and tools.py for reusable helper functions.
    * Design Decisions:
        * Followed a clean architecture with clear separation of responsibilities to improve scalability and readability.
        * Emphasized reusability with helper abstractions.
5. Testing
    * Implementation:
        * Created unit tests (test_core.py, test_tools.py) for all primary and helper functions using PyTest.
        * Covered normal operations, edge cases, and error handling.
    * Design Decisions:
        * Prioritized test-driven development (TDD) to ensure code reliability.
        * Mocked dependencies for isolated testing and ensured 100% coverage for core functionalities.
6. Logging and Error Handling
    * Implementation:
        * Added comprehensive logging to track errors and user inputs for debugging.
    * Design Decisions:
        * Followed platform standards for consistent and meaningful error reporting.
        * Prioritized user-friendly error messages to enhance user experience.
7. Extensibility and Maintainability
    * Implementation:
       * Used metadata files (metadata.json) to define input parameters for validation.
        * Documented all functions with type hints and docstrings.
    * Design Decisions:
        * Focused on future-proofing by using modular and extensible code structures.
       * Prioritized readability with consistent naming conventions and clear comments.

## How to Test

 Test Configurations
1. Ensure you have the necessary dependencies installed: pip install -r requirements.txt 
2. Run the tests: PYTHONPATH=./ pytest app/features/notes_generator/tests/test_core.py

Test Cases and Expected Outcomes
1. test_executor_text_input_valid:
    * Description: Tests the executor function with valid text input.
    * Expected Outcome: The function should return a status of "success". 
2. test_executor_file_input_valid:
    * Description: Tests the executor function with valid file input.
    * Expected Outcome: The function should return a status of "success". 
3. test_executor_invalid_file_type:
    * Description: Tests the executor function with an unsupported file type.
    * Expected Outcome: The function should raise a ValueError.
4. test_extract_text_from_file:
    * Description: Tests the extract_text_from_file function with a valid file.
    * Expected Outcome: The function should return the extracted text.
5. test_extract_text_from_url:
    * Description: Tests the extract_text_from_url function with a valid URL.
    * Expected Outcome: The function should return the extracted text.

## Unit Tests

1. test_executor_text_input_valid:
    * Tests the executor function with valid text input.
2. test_executor_file_input_valid:
    * Tests the executor function with valid file input.
3. test_executor_invalid_file_type:
    * Tests the executor function with an unsupported file type.
4. test_extract_text_from_file:
    * Tests the executor function with dummy.pdf file.
6. test_extract_text_from_url:
    * Tests the executor function with valid YouTube video ID

## Documentation Updates

N/A

- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.

- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

## Additional Information

I am look forward to your review and feedback

### CODE REVIEW

File: app/features/notes_generator/core.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            dict: Status and file path of the exported notes.
+            """
+    try:
+        logging.info("Notes generation started...")
+
+        # Validate that at least one of input_content or file_path is provided
+        if not (input_content or file_path):
+            raise ValueError("Either 'input_content' or 'file_path' must be provided.")
+        # Extract content based on input type
+        if input_type == 'text' and input_content:
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: Use an scalable approach for implementing these conditions. You can check this resource for more info. on that: https://www.linkedin.com/pulse/optimizing-10000-if-else-statements-strategies-best-practices-asim-ammgf/

File: app/features/notes_generator/core.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            content = input_content
+        elif input_type == 'file' and file_path:
+            # This is a check to validate file type before proceeding
+            if not file_path.endswith(('.txt', '.docx', '.pdf', '.csv')):
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: These are not all the current file types that our backend supports. You can check our `document_loaders.py` for more insights on that.

File: app/features/notes_generator/core.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    generate_notes,
+    export_notes
+)
+
+def executor(input_type: Literal['text', 'file', 'url'],
+             input_content: Optional[str] = None,
+             file_path: Optional[str] = None,
+             focus_topic: Optional[str] = None,
+             output_structure: Literal['bullet', 'paragraph', 'table'] = 'bullet',
+             export_format: Literal['txt', 'docx', 'pdf'] = 'txt') -> Dict[str, str]:
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: The standardize "export format" that the Marvel AI Backend supports is the JSON format for our UI client.

File: app/features/notes_generator/core.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            if not file_path.endswith(('.txt', '.docx', '.pdf', '.csv')):
+                raise ValueError(f"Unsupported file type for file: {file_path}")
+            content = extract_text_from_file(file_path)
+            # The below is to clean the extracted text
+             # Remove non-ASCII characters
+            content = re.sub(r'[^\x00-\x7F]+', '', content)
+            # Remove null bytes
+            content = content.replace('\x00', '')
+            # Remove non-printable characters
+            content = ''.join([char if char.isprintable() else '' for char in content])
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: This content pre-processing and validation can be done in an util function for keeping the Single Responsability principle.

File: app/features/notes_generator/core.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        elif input_type == 'url' and input_content:
+            content = extract_text_from_url(input_content)
+        else:
+            raise ValueError("Invalid input or missing content.")
+        # Generate notes
+        notes = generate_notes(content, focus_topic, output_structure)
+        # Export notes
+        exported_file_path = export_notes(notes, export_format)
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: This export is not required.

File: app/features/notes_generator/metadata.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/metadata.json]: Please, the `metadata.json` attributes must follow the same principle and structure as the other tools.

File: app/features/notes_generator/tests/test_core.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+from app.features.notes_generator.core import executor
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tests/test_core.py]: The testing structure must follow the same principle and structure from the other tools.

File: app/features/notes_generator/tests/test_tools.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import pytest
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tests/test_tools.py]: The `test_core.py` file actually tests the tool itself. This file is not required for not generating redundancy.

File: app/features/notes_generator/tools.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+from typing import Optional
+import docx
+from pdfminer.high_level import extract_text as extract_pdf
+from youtube_transcript_api import YouTubeTranscriptApi
+from docx import Document
+from fpdf import FPDF
+
+def extract_text_from_file(file_path: str) -> str:
+    """extract text based on file type."""
+    if file_path.endswith('.pdf'):
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py]: Use an scalable approach for implementing these conditions. You can check this resource for more info. on that: https://www.linkedin.com/pulse/optimizing-10000-if-else-statements-strategies-best-practices-asim-ammgf/

File: app/features/notes_generator/tools.py
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py:1]: Where is the LLM Pipeline or RAG approach used for the Notes Generation? I see util functions for transcribing content.

File: dummy.pdf
Comment (AaronSosaRamos) [inline dummy.pdf:1]: This file is not required.

File: outputs/notes_output.docx
Comment (AaronSosaRamos) [inline outputs/notes_output.docx:1]: This file is not required.

File: outputs/notes_output.pdf
Comment (AaronSosaRamos) [inline outputs/notes_output.pdf:1]: This file is not required.

File: outputs/notes_output.txt
Comment (AaronSosaRamos) [inline outputs/notes_output.txt:1]: This file is not required.

File: requirements.txt [hunk 10]
```diff
@@ -36,3 +36,11 @@ pydub
 ffmpeg-python
 speechrecognition
 google-cloud-speech
+
+
+pytesseract
```
Comment (AaronSosaRamos) [inline requirements.txt]: Why is pytesseract needed?

File: requirements.txt [hunk 11]
```diff
@@ -36,3 +36,11 @@ pydub
...
+pdfminer.six
```
Comment (AaronSosaRamos) [inline requirements.txt]: Why is pdfminer.six needed?

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Good work. Please, solve the given comments.

File: app/features/notes_generator/core.py (hunk 3 above)
Comment (Haryohmi) [reply app/features/notes_generator/core.py]: Acknowledged, Thank you for the feedback and clarity. I will update that.

File: app/features/notes_generator/core.py (hunk 2 above)
Comment (Haryohmi) [reply app/features/notes_generator/core.py]: That is right, thank you

File: app/features/notes_generator/core.py (hunk 4 above)
Comment (Haryohmi) [reply app/features/notes_generator/core.py]: Thank you, I have Refactored into clean_content and moved it to tool.py

File: app/features/notes_generator/core.py (hunk 5 above)
Comment (Haryohmi) [reply app/features/notes_generator/core.py]: Thank you. I have updated this.

File: app/features/notes_generator/tools.py (hunk 9 above)
Comment (Haryohmi) [reply app/features/notes_generator/tools.py]: Acknowledged. Thank you.

File: dummy.pdf
Comment (Haryohmi) [reply dummy.pdf:1]: Noted. Thank you.

File: outputs/notes_output.docx
Comment (Haryohmi) [reply outputs/notes_output.docx:1]: Noted. Thank you.
it was generated during test.

File: app/features/notes_generator/tools.py
Comment (Haryohmi) [reply app/features/notes_generator/tools.py:1]: Acknowledge. I will be implementing RAG approach for the Notes Generation.

**Review by Haryohmi — COMMENTED** [review]
I have updated and resolved all comments for your review.

File: .gitignore [hunk 12]
```diff
@@ -163,7 +163,6 @@ cython_debug/
 **/key.txt
 **/local-auth.json
 .DS_Store
-
```
Comment (AaronSosaRamos) [inline .gitignore]: Why is this change needed?

File: app/features/notes_generator/core.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    extract_text_from_file,
+    extract_text_from_url,
+    generate_notes_rag,
+    clean_content
+)
+
+# Dictionary to map input types to corresponding functions
+file_handler_map = {
+    'text': lambda content: clean_content(content),  # Clean content directly
+    'file': extract_text_from_file,
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: In the Marvel AI Backend, we are managing files with URLs and specifying the file type. Please, remove this for standardization purposes.

File: app/features/notes_generator/core.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    'url': extract_text_from_url
+}
+
+# List of supported file extensions
+SUPPORTED_FILE_TYPES = ['.txt', '.docx', '.pdf', '.csv', '.pptx', '.md', '.xml', '.xls', '.xlsx']
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: These are not all the supported file types from our Marvel AI Backend. You can check the `utils/document_loaders.py` file for knowing which file types are supported for our system.

File: app/features/notes_generator/core.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+def executor(input_type: Literal['text', 'file', 'url'],
+             input_content: Optional[str] = None,
+             file_path: Optional[str] = None,
+             focus_topic: Optional[str] = None,
+             output_structure: Literal['bullet', 'paragraph', 'table'] = 'bullet',
+             export_format: Literal['json'] = 'json') -> Dict[str, str]:
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: This is not needed

File: app/features/notes_generator/core.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+# List of supported file extensions
+SUPPORTED_FILE_TYPES = ['.txt', '.docx', '.pdf', '.csv', '.pptx', '.md', '.xml', '.xls', '.xlsx']
+
+def executor(input_type: Literal['text', 'file', 'url'],
+             input_content: Optional[str] = None,
+             file_path: Optional[str] = None,
+             focus_topic: Optional[str] = None,
+             output_structure: Literal['bullet', 'paragraph', 'table'] = 'bullet',
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: We need a `lang` variable here for multilingual purposes.

File: app/features/notes_generator/core.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+        # Validate the file type if file_path is provided
+        if file_path and not any(file_path.endswith(ext) for ext in SUPPORTED_FILE_TYPES):
+            raise ValueError(f"Unsupported file type for file: {file_path}")
+
+        # Extract content based on input type
+        if input_type in file_handler_map:
+            if input_type == 'text' and input_content:
+                content = file_handler_map[input_type](input_content)
+            elif input_type == 'file' and file_path:
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/core.py]: In the Marvel AI Backend, we are managing files with URLs and specifying the file type. Please, remove this for standardization purposes.

File: app/features/notes_generator/metadata.json [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    "name": "notes_generator",
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/metadata.json]: The `metadata.json` attributes are still not aligned similarly as the other tools.

File: app/features/notes_generator/tests/test_tools.py
Comment (AaronSosaRamos) [inline app/features/notes_generator/tests/test_tools.py:1]: These unit tests are not aligned with our guidelines. Please, check the `test_core.py` unit tests for standardization.

File: app/features/notes_generator/tools.py
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py:1]: All the cleaning functions are not needed since we have our `utils/document_loaders.py` that you can use for standradizing the `Document` class from LangChain.

File: app/features/notes_generator/tools.py [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+    if file_extension in file_handler_map:
+        handler = file_handler_map[file_extension]()  # Instantiate the handler
+        return handler(file_path)
+    else:
+        raise ValueError(f"Unsupported file type: {file_extension}")
+
+# Function to generate notes with RAGB from the extracted content
+def generate_notes_rag(content: str, focus_topic: Optional[str], structure: str) -> str:
+    # 1. Create embeddings and vector store from content (one-time or repeated)
+    embeddings = OpenAIEmbeddings(openai_api_key="OPEN_AI_API_KEY")
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py]: We don't use OpenAI in our codebase. We use Google Gemini. Please, align with that.

File: app/features/notes_generator/tools.py [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    # This can handle large content by splitting it into documents
+    docs = [content]  # If you split content into chunks, store them as multiple docs
+    vectorstore = FAISS.from_texts(docs, embeddings)
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py]: Why is the purpose of using FAISS as the vectorstore? In our codebase, we use Chroma

File: app/features/notes_generator/tools.py [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+
+    # 2. Create a QA chain using your vectorstore
+    retriever = vectorstore.as_retriever(search_kwargs={"k": 3})
+    qa_chain = RetrievalQA.from_chain_type(
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py]: Why is `RetrievalQA` needed? This class is deprecated since LangChain 0.1.17: https://python.langchain.com/api_reference/langchain/chains/langchain.chains.retrieval_qa.base.RetrievalQA.html

File: app/features/notes_generator/tools.py [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+        llm=ChatOpenAI(temperature=0.7, openai_api_key="OPEN_AI_API_KEY"),
```
Comment (AaronSosaRamos) [inline app/features/notes_generator/tools.py]: We don't use OpenAI in our codebase. We use Google Gemini. Please, align with that.

File: requirements.txt [hunk 23]
```diff
@@ -17,6 +17,7 @@ pypdf
 fpdf
 youtube-transcript-api
 pytube
+openai
```
Comment (AaronSosaRamos) [inline requirements.txt]: We don't use OpenAI in our codebase. We use Google Gemini. Please, align with that.

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Please, solve the given comments.
For better guidance in standardization, you can check how the Lesson Plan Generator was made: https://github.com/marvelai-org/marvel-ai-backend/tree/Develop/app/tools/lesson_plan_generator

File: app/features/notes_generator/metadata.json (hunk 18 above)
Comment (Haryohmi) [reply app/features/notes_generator/metadata.json]: Updated to match the style from lesson_plan_generator

File: requirements.txt (hunk 23 above)
Comment (Haryohmi) [reply requirements.txt]: Resolved.

File: app/features/notes_generator/tools.py (hunk 22 above)
Comment (Haryohmi) [reply app/features/notes_generator/tools.py]: Resolved

File: app/features/notes_generator/tools.py (hunk 21 above)
Comment (Haryohmi) [reply app/features/notes_generator/tools.py]: Thank you

File: app/features/notes_generator/tools.py (hunk 20 above)
Comment (Haryohmi) [reply app/features/notes_generator/tools.py]: Resolved to implement Chroma

File: app/features/notes_generator/tools.py (hunk 19 above)
Comment (Haryohmi) [reply app/features/notes_generator/tools.py]: Resolved
Thank you

**Comment by AaronSosaRamos** [comment]
Hey @Haryohmi good work so far. Please, check this guidelines about how to create your own Marvel AI Tool for better enhancements: https://lab.realityai.tech/posts/how-to-create-your-marvel-ai-tool

**Comment by AaronSosaRamos** [comment]
Hey @Haryohmi, please let me know when you update your Notes Generator proposal in base of the given guidelines. So far from that, good work in the approach.

**Comment by Haryohmi** [comment]
@AaronSosaRamos  Thank you I will do that, I am working on making sure it all aligns with the guidelines and standards

**Comment by Haryohmi** [comment]
@AaronSosaRamos  I have now updated the Notes Generator proposal in base of the given guidelines. Please let me know if any further changes is needed.

File: app/tools/notes_generator/core.py [hunk 24]
```diff
@@ -0,0 +1,76 @@
+from typing import Optional, Dict
+from app.api.error_utilities import FileHandlerError
+from app.utils.document_loaders import get_docs
+from app.tools.notes_generator.tools import NotesGeneratorPipeline
+from app.services.logger import setup_logger
+
+logger = setup_logger(__name__)
+
+def executor(
+    doc_url: Optional[str] = None,
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py]: Almost all the inputs of the executor are not Optional. If there is an input which an either be text or a file, that can be Optional

File: app/tools/notes_generator/core.py [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+        if doc_url and doc_type:
+            if not isinstance(doc_type, str) or doc_type not in SUPPORTED_DOC_TYPES:
+                raise ValueError(f"Unsupported doc_type: {doc_type}. Must be one of {SUPPORTED_DOC_TYPES}.")
+            try:
+                docs = get_docs(doc_url, doc_type, verbose=True)  # get_docs function from document_loaders.py
+            except FileHandlerError as fnf:
+                logger.error(f"Failed to load docs from {doc_url}: {str(e)}")
+                raise ValueError(f"Failed to load docs from {doc_url} - {str(e)}")
+
+        # 2. If user gave raw text, we can wrap it into a single-document list
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py]: It is not required to use a single-document list if that is a raw text, just use the text.

File: app/tools/notes_generator/core.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+        if text_content:
+            from langchain_core.documents import Document
+            docs.append(Document(page_content=text_content))
+
+        # 3. Run the pipeline to generate notes
+        pipeline = NotesGeneratorPipeline(
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py:55]: The pipeline must use all the given inputs in the executor

File: requirements.txt (hunk 23 above)
Comment (AaronSosaRamos) [reply requirements.txt]: Great @Haryohmi, please remove the `openai` dependency

File: app/tools/notes_generator/metadata.json
Comment (AaronSosaRamos) [inline app/tools/notes_generator/metadata.json:1]: Good approach with the metadata.json. Just align them with the required inputs that you can see in the Design section in #129

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Hey @Haryohmi, good work on the feature. You have improved a lot with the given guidelines. Please, solve the given comments. Thank you,

**Comment by Haryohmi** [comment]
@AaronSosaRamos Thank you, I have now updated the executor, pipeline, requirements and metadata.json . I am a bit skeptical about if "label": "File/URL Type" is necessary in the metadata.json. Please let me know if any further adjustment is required. Thank you once again

File: app/tools/notes_generator/core.py [hunk 27]
```diff
@@ -0,0 +1,78 @@
+from typing import Dict
+from app.api.error_utilities import FileHandlerError
+from app.utils.document_loaders import get_docs
+from app.tools.notes_generator.tools import NotesGeneratorPipeline
+from app.services.logger import setup_logger
+
+logger = setup_logger(__name__)
+
+def executor(
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py:9]: The args. must be `topic, page_layout, text, text_file_url, text_file_type`

File: app/tools/notes_generator/core.py [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+        doc_url (str): URL for the document to load.
+        doc_type (str): The type of document (pdf, docx, csv, etc.).
+        text_content (str): Plain text input if the user wants to pass text directly.
+        focus_topic (str): Focus or topic for the notes generation.
+        lang (str): Language code (e.g. 'en', 'es', 'fr', etc.) to guide multilingual generation.
+
+    Returns:
+        Dict[str, str]: Contains the status and generated notes.
+    """
+    SUPPORTED_DOC_TYPES = {"pdf","docx","csv","txt","md","pptx","xls","xlsx","gpdf"}
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py]: This is not required as we have our document loaders, please remove that.

File: app/tools/notes_generator/core.py [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+                logger.error(f"Failed to load docs from {doc_url}: {str(e)}")
+                raise ValueError(f"Failed to load docs from {doc_url} - {str(e)}")
+
+        # 2. If user gave raw text
+        if text_content:
+            text_content=text_content
+
+        # 3. Run the pipeline with all input to generate notes
+        pipeline = NotesGeneratorPipeline(
+            doc_url=doc_url or "",
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/core.py]: The args. must be `topic, page_layout, text, text_file_url, text_file_type`

File: app/tools/notes_generator/metadata.json
Comment (AaronSosaRamos) [inline app/tools/notes_generator/metadata.json:1]: The `name` and `description` of the Tool are actually located in the `tools_config.json` and in the `README.md` respectively. The label from each input is actually the description, and we can assign if each input can be required by using the `Optional` class from the Pydantic schemas. For example, text, text_file_url and text_file_type can be Optional. Additionally, these args. must be equal as those which are located in the `executor` function from the `core.py`.

File: app/tools/notes_generator/tests/test_core.py
Comment (AaronSosaRamos) [inline app/tools/notes_generator/tests/test_core.py:1]: Good approach. Please, update this with the given feedback.

File: app/tools/notes_generator/tools.py [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+
+class NotesGeneratorPipeline:
+    """
+    A pipeline for generating notes using a retrieval-based approach 
+    plus a generative LLM (Google Gemini).
+    """
+
+    def __init__(
+        self,
+        doc_url: str = "",
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/tools.py]: The args. must be `topic, page_layout, text, text_file_url, text_file_type`

File: app/tools/notes_generator/tools.py [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+            prompt_template = f"Summarize the content focusing on '{self.focus_topic}'. Language: {self.lang}"
+            # I believe we can add a more advanced multi-step retrieval,
+
+            doc_summary = summarize_chain.run({"input_documents": docs, "question": prompt_template})
+
+        # If we also have text_content, or only text_content, do a direct LLM call
+        text_summary = ""
+        if self.text_content and not docs:
+            # Summarize raw text alone
+            text_summary = self.model.invoke(
```
Comment (AaronSosaRamos) [inline app/tools/notes_generator/tools.py]: Why is text summarization required? How do you keep the context from the text for giving better responses? One suggestion is to have two different methods, one if there is a file_type and a file_url for obtaining a retriever, and the other method if there is just a text variable without using a vectorstore.

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Good work @Haryohmi, you are improving with the given feedback. 
Please, solve the given comments.
Thank you :)

**Comment by AaronSosaRamos** [comment]
> @AaronSosaRamos Thank you, I have now updated the executor, pipeline, requirements and metadata.json . I am a bit skeptical about if "label": "File/URL Type" is necessary in the metadata.json. Please let me know if any further adjustment is required. Thank you once again

Thank you for asking that @Haryohmi. We specify those args. because some tools can allow two different files to be loaded, so we can't generalize that for all the tools and that is why we specify that in the `metadata.json` file.

File: app/tools/notes_generator/tests/test_core.py
Comment (Haryohmi) [reply app/tools/notes_generator/tests/test_core.py:1]: Thank you @AaronSosaRamos

File: app/tools/notes_generator/core.py (hunk 28 above)
Comment (Haryohmi) [reply app/tools/notes_generator/core.py]: I have update every other part with the feedback but regarding this, I use it locally validate the file_type before calling get_docs. Because get_docs is not returning error for invalid file_types. 

Currently I handle invalid file as error if it is not a string which is not too right, Kindly review and advice. Thank you

**Comment by Haryohmi** [comment]
@AaronSosaRamos The comments have been resolved. I look forward to further review

File: .github/workflows/main.yml
Comment (stevenrayhinojosa-gmail-com) [inline .github/workflows/main.yml:1]: Summary of Fixes Needed:
Remove duplicate and conflicting name fields.
Eliminate duplicate branch definitions in on.push.
Correct job indentation and structure.
Simplify and deduplicate redundant steps (e.g., Google Cloud Auth, Cloud SDK setup).
Dynamically handle updates to app.yaml without overwriting existing content.
Fix environment variable syntax for proper resolution.
Add error handling for critical steps.

File: app/utils/document_loaders.py
Comment (stevenrayhinojosa-gmail-com) [inline app/utils/document_loaders.py:1]: this file is missing the documnet loaders, please update.

**Comment by stevenrayhinojosa-gmail-com** [comment]
@yunusj ready for your review
